mmengine/runner/_flexible_runner.py

```diff
@@ -0,0 +1,1660 @@
+                f'train_dataloader={train_dataloader}, '
+                f'train_cfg={train_cfg}, '
+                f'optim_wrapper={optim_wrapper}.')
+        self._train_dataloader = train_dataloader
+        self._train_loop = train_cfg
+
+        self.optim_wrapper: Optional[Union[OptimWrapper, dict]]
+        self.optim_wrapper = optim_wrapper
+
+        self._auto_scale_lr = auto_scale_lr
+
+        # If there is no need to adjust learning rate, momentum or other
+        # parameters of optimizer, param_scheduler can be None
+        if param_scheduler is not None and self.optim_wrapper is None:
+            raise ValueError(
+                'param_scheduler should be None when optim_wrapper is None, '
+                f'but got {param_scheduler}')
+
+        # Parse `param_scheduler` to a list or a dict. If `optim_wrapper` is a
+        # `dict` with single optimizer, parsed param_scheduler will be a
+        # list of parameter schedulers. If `optim_wrapper` is
+        # a `dict` with multiple optimizers, parsed `param_scheduler` will be
+        # dict with multiple list of parameter schedulers.
+        self._check_scheduler_cfg(param_scheduler)
+        self.param_schedulers = param_scheduler
+
+        val_related = [val_dataloader, val_cfg, val_evaluator]
+        if not (all(item is None
+                    for item in val_related) or all(item is not None
+                                                    for item in val_related)):
+            raise ValueError(
+                'val_dataloader, val_cfg, and val_evaluator should be either '
+                'all None or not None, but got '
+                f'val_dataloader={val_dataloader}, val_cfg={val_cfg}, '
+                f'val_evaluator={val_evaluator}')
+        self._val_dataloader = val_dataloader
+        self._val_loop = val_cfg
+        self._val_evaluator = val_evaluator
+
+        test_related = [test_dataloader, test_cfg, test_evaluator]
+        if not (all(item is None for item in test_related)
+                or all(item is not None for item in test_related)):
+            raise ValueError(
+                'test_dataloader, test_cfg, and test_evaluator should be '
+                'either all None or not None, but got '
+                f'test_dataloader={test_dataloader}, test_cfg={test_cfg}, '
+                f'test_evaluator={test_evaluator}')
+        self._test_dataloader = test_dataloader
+        self._test_loop = test_cfg
+        self._test_evaluator = test_evaluator
+
+        if not isinstance(compile, bool) and not isinstance(compile, dict):
+            raise TypeError(
+                f'compile should be a bool or dict, but got {type(compile)}')
+        self._compile = compile
+
+        self._load_from = load_from
+        self._resume = resume
+        # flag to mark whether checkpoint has been loaded or resumed
+        self._has_resumed = False
+
+        if launcher is None:
+            launcher = infer_launcher()
```

Comment:
How about inferring `launch` in `strategy`

Reply:
The launcher is needed for deciding which strategies will be used when `launcher` is None.

Reply:
Also do it in Strategy.